## Deep Learning-based Clustering Approaches for Bioinformatics
Codes and supplementary materials for our paper "Deep Learning-based Clustering Approaches for Bioinformatics" has been accepted for publication in [Briefings in Bioinformatics](https://academic.oup.com/bib) journal. This repo will be updated periodically. In particular, more complete Jupyter notebooks will be added and new approaches/paper will be listed. 

In this article, we reviewed deep learning-based approaches for cluster analysis, including network training, representation learning, parameter optimization, and formulating clustering quality metrics. We also discussed how representation learning based on different autoencoder architectures (e.g., vanilla, variational, LSTM, and convolutional) can be more effective than machine learning-based approaches (e.g., PCA) in different scenarios, e.g., bio-imaging, gene expression clustering, and clustering biomedical texts. 

The article is in the press and soon will be published online with open access. 

### Deep learning-based unsupervised/clustering methods, link to papers, and codes
|    Title    |  Article  |  Conference/Journal | Code |
|  :---------  | :------:  | :------: | :------: |
| Deep clustering with convolutional autoencoders (DCEC) | [Link](https://xifengguo.github.io/papers/ICONIP17-DCEC.pdf) | ICONIP'2017 |  [GitHub](https://github.com/XifengGuo/DCEC) |
| Unsupervised Data Augmentation for Consistency Training (UDA) | [Link](https://arxiv.org/pdf/1904.12848.pdf) | Arxiv'2019 |  [GitHub](https://github.com/google-research/uda) |
| Deep Clustering via joint convolutional autoencoder embedding and relative entropy minimization (DEPICT) | [Link](https://arxiv.org/pdf/1704.06327.pdf) | ICCV'2017 |  [GitHub](https://github.com/herandy/DEPICT) |
| Discriminatively Boosted Clustering (DBC) | [Link](https://arxiv.org/pdf/1703.07980.pdf) | Arxiv'2017 | N/A|
| Variational Deep Embedding (VADE) | [Link](https://arxiv.org/pdf/1611.05148.pdf) | IJCAI'2017 |  [GitHub](https://github.com/slim1017/VaDE) |
| Convolutional Embedded Networks (CEN)} | [Link](https://arxiv.org/pdf/1805.12218.pdf) | Arxiv'2018 |  [GitHub](https://github.com/rezacsedu/Convolutional-embedded-networks) |
| Deep Subspace Clustering Networks (DSC-Nets) | [Link](http://papers.nips.cc/paper/6608-deep-subspace-clustering-networks.pdf) | NIPS'2017 |[GitHub](https://github.com/panji1990/Deep-subspace-clustering-networks) |
| Graph Clustering with Dynamic Embedding (GRACE) | [Link](https://arxiv.org/pdf/1712.08249.pdf) | Arxiv'2017 | N/A |
| Deep Unsupervised Clustering Using Mixture of Autoencoders (MIXAE) | [Link](https://arxiv.org/pdf/1712.07788.pdf) | Arxiv'2017 |N/A |
|  Deep Embedded Clustering (DEC) | [Link](http://proceedings.mlr.press/v48/xieb16.pdf) |   ICML'2016  | [GitHub](https://github.com/piiswrong/dec) |
|  A Survey of Clustering With Deep Learning: From the Perspective of Network Architecture | [Link](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8412085)  |  IEEE ACCESS 2018  |  |
|  GEMSEC: Graph Embedding with Self Clustering  | [Link](https://arxiv.org/pdf/1802.03997.pdf) | Arxiv,2018  | [GitHub](https://github.com/benedekrozemberczki/GEMSEC) |
|  Clustering with Deep Learning: Taxonomy and New Methods | [Link](https://arxiv.org/pdf/1801.07648.pdf)  |  Arxiv, 2018 |  [GitHub](https://github.com/elieJalbout/Clustering-with-Deep-learning)  |
| Deep Continuous Clustering (DCC) | [Link](https://arxiv.org/pdf/1803.01449.pdf) | Arxiv, 2018 | [GitHub](https://github.com/shahsohil/DCC) |
| Deep Clustering with Convolutional Autoencoders (DCEC) | [Link](https://xifengguo.github.io/papers/ICONIP17-DCEC.pdf) | ICONIP'2018 | [GitHub](https://github.com/XifengGuo/DCEC) |
| SpectralNet: Spectral Clustering Using Deep Neural Networks | [Link](https://openreview.net/pdf?id=HJ_aoCyRZ) | ICLR'2018 | [GitHub](https://github.com/KlugerLab/SpectralNet) |
| Subspace clustering using a low-rank constrained autoencoder (LRAE) | [Link](https://www.sciencedirect.com/science/article/pii/S0020025517309659) | Information Sciences'2018 | N/A|
| Clustering-driven Deep Embedding with Pairwise Constraints (CPAC) | [Link](https://arxiv.org/pdf/1803.08457.pdf) | Arxiv'2018 |  [GitHub](https://github.com/sharonFogel/CPAC) |
|  Towards K-means-friendly Spaces: Simultaneous Deep Learning and Clustering | [Link](https://arxiv.org/pdf/1610.04794.pdf)  |  PMLR'2017  |  N/A  |
| Deep Unsupervised Clustering With Gaussian Mixture Variational AutoEncoders (GMVAE) | [Link](https://arxiv.org/pdf/1611.02648.pdf)| ICLR'2017 | [GitHub](https://github.com/Nat-D/GMVAE)| N/A |
|  Is Simple Better?: Revisiting Simple Generative Models for Unsupervised Clustering | [Link](https://ic.unicamp.br/~adin/downloads/pubs/AriasFigueroa2017a.pdf)  |  NIPS'2017 Workshop  |  [GitHub](https://github.com/jariasf/clustering-nips-2017)  |
|      Imporved Deep Embedding Clustering (IDEC) | [Link](https://www.ijcai.org/proceedings/2017/0243.pdf)   |   IJCAI'2017  |  [GitHub](https://github.com/XifengGuo/IDEC) |
| Deep Clustering Network (DCN) | [Link](https://arxiv.org/pdf/1610.04794v1.pdf)  | Arxiv'2016 | [GitHub](https://github.com/boyangumn/DCN-New) | N/A|
| Joint Unsupervised Learning of Deep Representations and Image Clustering (JULE) | [Link](https://arxiv.org/pdf/1604.03628.pdf) | CVPR'2016 |   [GitHub](https://github.com/jwyang/JULE.torch) |
| Deep Embedding Network for Clustering (DEN) | [Link](https://ieeexplore.ieee.org/document/6976982/) | ICPR'2014 | N/A|
| Auto-encoder Based Data Clustering (ABDC) | [Link](http://nlpr-web.ia.ac.cn/english/irds/People/lwang/M-MCG_EN/Publications/2013/CFS2013CIARP.pdf) | CIARP'2013 |  [GitHub](https://github.com/KellerJordan/Autoencoder-Clustering) |
| Learning Deep Representations for Graph Clustering | [Link](https://www.aaai.org/ocs/index.php/AAAI/AAAI14/paper/viewFile/8527/8571) |  AAAI'2014 | [GitHub](https://github.com/quinngroup/deep-representations-clustering) |

## Running provided Jupyter notebooks
To run the examples interactively, you need to install some Python modules and libraries. 

* Python 3
* Scikit-learn
* Keras 
* TensorFlow.

For the Jupyter notebook, git it from this [Link](https://jupyter.readthedocs.io/en/latest/install.html) and install it on your machine. Then clone this repo using following command, given that you have already installed the `git`: 

```
git clone https://github.com/rezacsedu/Deep-learning-for-clustering-in-bioinformatics.git
```
Alternatively, install all the required libraries by issuing the following command: 
```
 cd Deep-learning-for-clustering-in-bioinformatics
 pip3 install -r requirements.txt
 cd Notebboks
 ```
Then start Jupyter notebbok by issuing the following command: 
```
jupyter notebook
```
In the opened browser, go to Jupyter tab and window open the notebook. 
```
LSTM_AE_Text_Clustering.ipynb
```
If you want to skip the training, soon we'll provide the pre-trained weights, which you can restore and start fine-tuning. Happy coding! Leave a comment if you have any question.

### Acknowledgement
The ClusteringLayer class and the target_distribution function are based on DEC from https://github.com/XifengGuo/DCEC/blob/master/DCEC.py by Xifeng Guo

### Citation request
If you use the code of this repository in your research, please consider citing the folowing papers:

    @inproceedings{karimBIB2020,
        title={Deep Learning-based Clustering Approaches for Bioinformatics},
        author=Md. Rezaul Karim, Oya Beyan, Achille Zappa, Ivan G. Costa, Dietrich Rebholz-Schuhmann, Michael Cochez, and Stefan Decker},
        journal={Briefings in Bioinformatics},
        year={2020}
    }

### Contributing
If you find more related work, which are not listed here, please create a PR or sugest by filing issues. Your contribution will be highly appreciated. For any questions, feel free to open an issue or contact at rezaul.karim@rwth-aachen.de. 
